### PR TITLE
Add routes and model changes for group chat #47

### PR DESCRIPTION
**Pull request by leiwang758**
### What this PR does (required):
- Add routes for creating a new group chat and add a user to a group chat
-  Each time a new user is added to the group, the message will automatically be translated and added to the translation list (if the language has not yet been supported)
- Change conversation model to include language supports
- Change postMessage route accordingly (looking for language supports directly from the conversation model instead of from each user)

### Screenshots / Videos (required):
- creating new group chat
image
 - posting messages (only translated to french so far)
image
 - After adding new user(s) (also translate all previous message to russian)
 - Now support adding a list of users as well
 Before adding
image
post request
image
After adding
image

### CODE REVIEW

File: server/controllers/conversation.js [hunk 1]
```diff
@@ -25,31 +27,125 @@ exports.getUserConversations = asyncHandler(async (req, res) => {
...
 
+  const currentUser = await User.findById(userId);
+  const theOtherUser = await User.findById(otherUserId);
   const conversationExists = await Conversation.findOne({
     users: [otherUserId, userId].sort(),
   });
 
   if (conversationExists) {
-    return res.status(200).json({ id: conversationExists._id });
+    return res.status(200).json({ message: "Conversation already exists" });
```
Comment (dproc96) [inline server/controllers/conversation.js:42]:
```suggestion
    return res.status(400).json({ message: "Conversation already exists" });
```
I realize this was already a 200 in the code before this PR, bbut probably better for it to be a 400

File: server/controllers/conversation.js [hunk 2]
```diff
@@ -25,31 +27,125 @@ exports.getUserConversations = asyncHandler(async (req, res) => {
...
+// create a new group chat with multiple users
+exports.postGroupChat = asyncHandler(async (req, res) => {
+  const userId = req.user.id;
+  const users = req.body.users;
+  const languages = await Promise.all(
+    users.map(async (uid) => {
+      const user = await User.findById(uid);
+      return user.primaryLanguage;
+    })
+  );
```
Comment (dproc96) [inline server/controllers/conversation.js:73]: Very nice use of Promises here 👍

**Comment by leiwang758** [comment]
Add filter on both users and languages so that only users that are not already in the group are added and only the languages that are not already supported will be added, also remove duplicates of newly added languages
```javascript
const filteredUsers = usersToAdd.filter(
      (user) => !conversation.users.includes(user)
    );
const filteredLanguages = languages.filter(
      (lang, index) =>
        !conversation.languages.includes(lang) &&
        languages.indexOf(lang) === index
    );
```

File: server/controllers/conversation.js [hunk 3]
```diff
@@ -34,22 +37,134 @@ exports.postUserConversation = asyncHandler(async (req, res) => {
...
+    // Check if all the new users already exist in the group
+    // and filter out the users that are already in the group
+    const filteredUsers = usersToAdd.filter(
+      (user) => !conversation.users.includes(user)
+    );
+    if (!filteredUsers?.length) {
+      return res.status(400).json({
+        message: "No user(s) to add in this group",
+      });
+    }
```
Comment (dproc96) [inline server/controllers/conversation.js:109]: This isn't the _most_ efficient way to accomplish this, but for this use case I think it's fine. If we're thinking of it from Big O notation (ex. O(n)) the n being users and languages is never going to be so large as to be a problem (like 100+ users in one group chat) so fine to leave like this.

**Review by dproc96 — APPROVED** [review]
Looks good to me!

**Pull request merged**